### DISCUSSION TITLE
Custom directional emitter with position

### DISCUSSION THREAD

**Discussion by duongnlt** [mitsuba-renderer/mitsuba3#966] [Q&A]
Hi team,

I am trying to create a scene with a rectangle polarizer placed in front of the sensor, an object and a directional emitter. The problem is in some light directions, the rectangle polarizer casts a shadow on my object which is not the outcome I expect. So I am wondering if it is possible to create a custom directional emitter that has a position property (like a point emitter) so that I can place it between the object and the polarizer. Or if you come up with another solution, I would really appreciate it.

Thank you in advance.

**Comment by njroussel** [comment]
Hi @duongnlt 

The rectangle should not be casting a shadow. Are you using the `volpath` integrator? As suggested in [the documentation](https://mitsuba.readthedocs.io/en/latest/src/generated/plugins_bsdfs.html#linear-polarizer-material-polarizer).

**Reply by duongnlt** [reply]
Hi @njroussel

I already used `volpath` as my integrator, here is the result
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/44237568/124ba144-85ab-4234-8e36-5745d038d54e)

**Reply by duongnlt** [reply]
Here is how I set up the scene
```python
import mitsuba as mi
import drjit as dr
import numpy as np

mi.set_variant('scalar_spectral_polarized')
from mitsuba import ScalarTransform4f as T

sensor = mi.load_dict({
    'type': 'perspective',
    'fov': 41,
    # 'fov_axis': 'larger',
    'to_world': T.translate([0,0,-7]),
    'sampler': {
        'type': 'multijitter',
        'sample_count': 64
    },
    'film': {
        'type': 'hdrfilm',
        'width': 256,
        'height': 256,
        'pixel_format': 'rgb',
        'rfilter': {
            'type': 'gaussian'
        }
    }
        
})
integrator = mi.load_dict({
                'type': 'volpath',
    
            },)
bunny = mi.load_dict({
            'type': 'obj',
            'filename': 'pot1.obj',
                
            'to_world': T.translate([-0.2,-0.5,2.2]).scale(0.02).rotate(axis=[1,0,0], angle=-120).rotate(axis=[0,0,1], angle=90)            
    
    
            'bsdf': {
                    'type': 'measured_polarized',
                    'filename': '../scene/BRDF/2_white_billiard_inpainted.pbsdf',
                    'alpha_sample': 0.1
            },
})

polarizer = mi.load_dict({
    'type': 'rectangle',
    'bsdf': {
        'type': 'polarizer',
        'theta': {
            'type': 'spectrum',
            'value': 0
        },
    },

    'to_world': T.translate([0,0,-6.5]).scale(0.2).rotate(axis=[0, 1, 0], angle=-180)
})

scene = mi.load_dict({
            'type':'scene',
            'integrator': integrator,
            'sensor': sensor,
            'object': bunny,
            'polarizer': polarizer,
            
            'emitter': {
                'type': 'directional',
                'irradiance': {
                    'type': 'rgb',
                    'value': [0.5368, 0.7229, 0.9031],
                }
        },
    
})
```

**Reply by njroussel** [reply]
Ahh, I see. I believe this is because the polarizer is effectively applied twice to certain rays/paths.

> that I can place it between the object and the polarizer

This would indeed be the solution. I think you could actually implement this already with a `directionalarea` emitter attached to a `rectangle`. You would need to apply a `null` BSDF to that emitter, such that you can see "through" it.
I think this should work.

**Reply by duongnlt** [reply]
> This would indeed be the solution. I think you could actually implement this already with a `directionalarea` emitter attached to a `rectangle`. You would need to apply a `null` BSDF to that emitter, such that you can see "through" it. I think this should work.

One problem is that a `directionalarea` emitter only works when using `ptracer` integrator, but the `null` BSDF needs `volpath` integrator. Both those integrators can not be set together.

**Reply by njroussel** [reply]
Oh indeed, I hadn't thought about that.

I don't we currently have a way to handle this then. You might be able to implement a custom directional emitter that starts on a specific plane.

**Comment by duongnlt** [comment]
> Ahh, I see. I believe this is because the polarizer is effectively applied twice to certain rays/paths.
> 
> > that I can place it between the object and the polarizer
> 
> This would indeed be the solution. I think you could actually implement this already with a `directionalarea` emitter attached to a `rectangle`. You would need to apply a `null` BSDF to that emitter, such that you can see "through" it. I think this should work.

@njroussel Follow your approach, I change my `scene` to
```python
scene = mi.load_dict({
            'type':'scene',
            'integrator': integrator,
            'sensor': sensor,
            'object': bunny,
            # 'polarizer': polarizer,
            
            'source': {
                'type': 'rectangle',
                'bsdf': {
                    'type': 'null'
                },
                'to_world': T.translate([0,0,-3]).rotate(axis=[1,0,0], angle=180),
                'emitter': {
                    'type': 'directionalarea',
                    'radiance': {
                        'type': 'rgb',
                        'value': 1.0,
                    }
                }
        },    
})
```
but the result is only a black image. The light only appears when I switch the `source` shape to `sphere`.
